### PR TITLE
SON object and operations

### PR DESCRIPTION
**Pull request by oxarbitrage**
Replaces https://github.com/peerplays-network/peerplays/pull/48

I toke the initial work from @pixelplex , made some naming changes, added the update operation, added hardfork guards and other small changes.

The next step should be to review and port the code from https://github.com/peerplays-network/peerplays/pull/49/commits/b3786b3626f70f8e392335e1f9c3ad428986b3bf into here in one side, this will allow to vote for this objects.

In the other side we can follow with creating the needed cli_wallet commands described in internal document. In order to do this we are going to need the cli_wallet tests from develop.

### CODE REVIEW

**Comment by obucinac** [comment]
Building error, SONs-base + 144

peerplays/peerplays/libraries/chain/proposal_evaluator.cpp:144:32: error: ‘HARDFORK_SON_TIME’ was not declared in this scope
       FC_ASSERT( block_time >= HARDFORK_SON_TIME, "son_create_operation not allowed yet!" );

**Comment by oxarbitrage** [comment]
my apologies, a file was missing.

**Comment by obucinac** [comment]
Regarding vote_id_type::son

database_api.cpp
vector<variant> database_api_impl::lookup_vote_ids( const vector<vote_id_type>& votes )const

There is a switch inside this function, and it does not have vote_id_type::son as one of the cases.

Should it be there?

**Comment by obucinac** [comment]
One more question about voting... With this PR, is SON voting functional, and how we can verify this? Can you provide a curl command or anything else?

**Comment by oxarbitrage** [comment]
This pull request is for this document: https://peerplays.atlassian.net/wiki/spaces/PIX/pages/333971489/SON+Objects+and+Operators
So, regarding to your question, we cant say voting for son is functional with this pull request.

The voting is described here as far as i know: https://peerplays.atlassian.net/wiki/spaces/PIX/pages/342261826/SON+Voting+LLD

And there is an attempt of implementation where some parts will be maybe reusable at: https://github.com/peerplays-network/peerplays/pull/49

In regards to the `lookup_vote_ids`, i added code for it, nice find.

**Comment by oxarbitrage** [comment]
IMHO, the document about voting is confusing: https://peerplays.atlassian.net/wiki/spaces/PIX/pages/342261826/SON+Voting+LLD

It introduce new fields to the son object and new fields and functionality(users must have a vesting balance to register)to the create operation in compassion with the former document, it introduces the delete operation again, it mentions sidechain diagrams that are unrelated. 
The modifications into the operations and objects should be moved into https://peerplays.atlassian.net/wiki/spaces/PIX/pages/333971489/SON+Objects+and+Operators where they belong. The unrelated sidechain things should be removed and placed in somewhere else. The voting document should specify only how son members can became active or inactive by voting(very similar to witness functionality) and nothing more.

**Comment by bobinson** [comment]
@satyakoneru  : please look into the comments on LLDs here.

File: tests/tests/son_operations_tests.cpp [hunk 1]
```diff
@@ -8,134 +8,134 @@
...
+
+   // create deposit vesting
+   vesting_balance_id_type deposit;
    {
-      trx_state = ptr_trx_state.get();
+      vesting_balance_create_operation op;
+      op.creator = alice_id;
+      op.owner = alice_id;
+      op.amount = asset(10);
+      op.balance_type = vesting_balance_type::unspecified;
```
Comment (bobinson) [inline tests/tests/son_operations_tests.cpp]: @oxarbitrage  : we have `normal` and `gpos` now

Comment (oxarbitrage) [reply tests/tests/son_operations_tests.cpp]: not in this branch: https://github.com/peerplays-network/peerplays/blob/feature/SONs-base/libraries/chain/include/graphene/chain/vesting_balance_object.hpp#L149

**Comment by bobinson** [comment]
Yea : we can revise this later

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by obucinac** [comment]
Quick question... Is this ready for a merge? Does it have all thats planned?

**Comment by oxarbitrage** [comment]
I had been out the last week but i think there was 1 thing missing here that was the vesting balance you need to have locked to became a son member. Needs to be tied to the create operation. Besides that the main things are all done. I will review the documents in case anything was changed or added and will add that functionality to merge if you agree. Voting should be done in a new pull request.

**Comment by obucinac** [comment]
I agree. Please finish vesting balance as soon as possible. Thanks

**Comment by oxarbitrage** [comment]
Most of the code in here was already merged to `feature/SONs-base` at https://github.com/peerplays-network/peerplays/pull/160 and the rest has now its own pr at https://github.com/peerplays-network/peerplays/pull/183

Closing this one as it is not needed anymore.

**Pull request closed without merging**